### DISCUSSION TITLE
How to confirm if limiter+redis is working?

### DISCUSSION THREAD

**Discussion by sbstratos79** [searxng/searxng#1114] [Q&A] [answered]
Is there a way to confirm if it's working as intended? I have setup my instance with limiter+redis and everything seems to be working fine. On the surface at least. I couldn't find a way to test if limiter is working or not though.

**Comment by mrpaulblack** [comment, marked as answer]
You can do a Curl request. The limiter currently is there to rate limited and protect searches. So for example this will be blocked by the limiter (since curl is a bot in its current impl.):
```
curl https://paulgo.io/search?q=time -i
```

You can change the domain to your instance and test it yourself. If you get something like this it works correctly:
```
HTTP/2 429 
content-security-policy: upgrade-insecure-requests; default-src 'none'; script-src 'self'; style-src 'self' 'unsafe-inline'; form-action 'self' https://github.com; font-src 'self'; frame-ancestors 'self'; base-uri 'self'; connect-src 'self' https://overpass-api.de; img-src 'self' data: https://*.tile.openstreetmap.org; frame-src https://www.youtube-nocookie.com https://player.vimeo.com https://www.dailymotion.com https://www.deezer.com https://www.mixcloud.com https://w.soundcloud.com https://embed.spotify.com
content-type: text/html; charset=utf-8
permissions-policy: accelerometer=();ambient-light-sensor=(); autoplay=();camera=();encrypted-media=();focus-without-user-activation=(); geolocation=();gyroscope=();magnetometer=();microphone=();midi=();payment=();picture-in-picture=(); speaker=();sync-xhr=();usb=();vr=()
referrer-policy: no-referrer
server-timing: total;dur=1.73, render;dur=0
strict-transport-security: max-age=63113904; includeSubDomains; preload
vary: Accept-Encoding
x-content-type-options: nosniff
x-download-options: noopen
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
content-length: 0
date: Mon, 18 Apr 2022 16:05:51 GMT
```

(Important is the HTTP 429 rate limit response)

Does this answer your question?

**Reply by sbstratos79** [reply]
Thanks for the answer. I tested it and it doesn't seem to be working for me. Got an HTTP 200 along with the complete html document as response.
I have added `limiter: true` under `server` settings and `url: unix:///usr/local/searxng-redis/run/redis.sock?db=0
` under `redis` settings in settings.yml. 
I also executed these commands after installing searxng. Am I missing some steps?
```
$ ./manage redis.build
$ sudo -H ./manage redis.install
$ sudo -H ./manage redis.addgrp "${USER}"
``` 

EDIT: I did some further testing under searxng shell, and I get this error for `redisdb.init()`:
```
2022-04-18 16:48:29,926 ERROR:searx.shared.redis: can't connet redis DB ...
2022-04-18 16:48:29,927 ERROR:searx.shared.redis:   Error 13 connecting to unix socket: /usr/local/searxng-redis/run/redis.sock. Permission denied.
False
```
Could you please tell me why I'm getting this error?

**Reply by mrpaulblack** [reply]
@return42 maybe a different user? A dirty fix could be to just change the permission on the socket file to like 755, but not recommended IMO.

**Reply by sbstratos79** [reply]
> @return42 maybe a different user? A dirty fix could be to just change the permission on the socket file to like 755, but not recommended IMO.

Thanks. Changing `unixsocketperm` value to 755 in `/usr/local/searxng-redis/redis.conf` did not help. I opened an issue here and somebody helped me solve the problem. `searx` needed to be added to `searx-redis` group. 
This also fixed the limiter configuration. Now it is properly blocking curl requests like in the example you provided.

**Reply by tiekoetter** [reply]
To elaborate on this: The ``searx`` user (the user running the actual SearXNG instance) needs access to the redis group (If you install redis via a package manager it is ``redis``. If you install via the scripts it is ``searxng-redis``).

**Reply by sbstratos79** [reply]
> To elaborate on this: The `searx` user (the user running the actual SearXNG instance) needs access to the redis group (If you install redis via a package manager it is `redis`. If you install via the scripts it is `searxng-redis`).

This should be added to the docs, right? Not adding `searx` to the appropriate group will cause the limiter etc. to not work without the user being even aware of the problem.

**Reply by Kopernikus1979** [reply]
I'm getting "HTTP/1.1 429 TOO MANY REQUESTS" so Redis/Limiter plugin is working?
Searxng and Redis both running as a docker container.

**Reply by mrpaulblack** [reply]
Yeah if you get a HTTP 429 response with curl than the limiter plugin in SearXNG with redis is working :+1:
